test/common/upstream/load_balancer_impl_test.cc

```diff
@@ -2880,6 +2881,96 @@ TEST_P(LeastRequestLoadBalancerTest, PNC) {
+                              makeTestHost(info_, "tcp://127.0.0.1:82", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:83", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:84", simTime())};
+  hostSet().hosts_ = hostSet().healthy_hosts_;
+  hostSet().runCallbacks({}, {}); // Trigger callbacks. The added/removed lists are not relevant.
+
+  hostSet().healthy_hosts_[0]->stats().rq_active_.set(4);
+  hostSet().healthy_hosts_[1]->stats().rq_active_.set(3);
+  hostSet().healthy_hosts_[2]->stats().rq_active_.set(2);
+  hostSet().healthy_hosts_[3]->stats().rq_active_.set(1);
+  hostSet().healthy_hosts_[4]->stats().rq_active_.set(5);
+
+  envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest lr_lb_config;
+
+  // Enable FULL_SCAN on hosts.
+  lr_lb_config.set_selection_method(
+      envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest::FULL_SCAN);
+
+  LeastRequestLoadBalancer lb{priority_set_, nullptr, stats_,       runtime_,
+                              random_,       1,       lr_lb_config, simTime()};
+
+  // With FULL_SCAN we will always choose the host with least number of active requests.
+  EXPECT_EQ(hostSet().healthy_hosts_[3], lb.chooseHost(nullptr));
+}
+
+TEST_P(LeastRequestLoadBalancerTest, FullScanMultipleHostsWithLeastRequests) {
+  hostSet().healthy_hosts_ = {makeTestHost(info_, "tcp://127.0.0.1:80", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:81", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:82", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:83", simTime()),
+                              makeTestHost(info_, "tcp://127.0.0.1:84", simTime())};
+  hostSet().hosts_ = hostSet().healthy_hosts_;
+  hostSet().runCallbacks({}, {}); // Trigger callbacks. The added/removed lists are not relevant.
+
+  hostSet().healthy_hosts_[0]->stats().rq_active_.set(3);
+  hostSet().healthy_hosts_[1]->stats().rq_active_.set(3);
+  hostSet().healthy_hosts_[2]->stats().rq_active_.set(1);
+  hostSet().healthy_hosts_[3]->stats().rq_active_.set(1);
+  hostSet().healthy_hosts_[4]->stats().rq_active_.set(1);
+
+  envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest lr_lb_config;
+
+  // Enable FULL_SCAN on hosts.
+  lr_lb_config.set_selection_method(
+      envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest::FULL_SCAN);
+
+  auto random = Random::RandomGeneratorImpl();
+
+  LeastRequestLoadBalancer lb{priority_set_, nullptr, stats_,       runtime_,
+                              random,        1,       lr_lb_config, simTime()};
+
+  // Make 1 million selections. Then, check that the selection probability is
+  // approximately equal among the 3 hosts tied for least requests.
+  // Accept a +/-0.5% deviation from the expected selection probability (33.3..%).
```

Comment:
@tonya11en : With a few runs, the max deviation I saw was ~0.1% (1000 counts off from 333333). I figured 0.5% deviation was enough to avoid most flakes while ensuring the probability remains equally distributed.

Reply:
Yeah, this is good. You can make the tolerance 1% just to be safe.

You can also have bazel run the test in parallel many times by appending `--runs_per_test=1000` to your `bazel test` command.

Reply:
not sure im entirely following the signififance of tolerance and deviation here - but please dont introduce any CI flakes

if it fails at all when running a 1000 tests in parallel then i think the test needs a rethink

Reply:
got it - thanks for explanation

Reply:
No flakes in 100 tests:
```
jkirschner@HashiH0PW6D3:~/envoy$ bazel test --jobs=6 --runs_per_test=100 //test/common/upstream:load_balancer_impl_test
INFO: Analyzed target //test/common/upstream:load_balancer_impl_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
[7 / 101] 6 actions running
[19 / 101] 6 actions running
[31 / 101] 6 actions running
[31 / 101] 6 actions running
[36 / 101] 6 actions running
[49 / 101] 6 actions running
[49 / 101] 6 actions running
[85 / 101] 6 actions running
[91 / 101] 6 actions running
Target //test/common/upstream:load_balancer_impl_test up-to-date:
  bazel-bin/test/common/upstream/load_balancer_impl_test
INFO: Elapsed time: 1929.003s, Critical Path: 134.90s
INFO: 101 processes: 1 internal, 100 linux-sandbox.
INFO: Build completed successfully, 101 total actions

Executed 1 out of 1 test: 1 test passes.
```

When I tried to run 1000 tests, 48 of them timed out (took a long time to run, and my machine goes to sleep quickly if I'm briefly inactive). I'm not sure if the below output means all the tests either passed or timed out (no true failures). I'm not sure how feasibly I can run 1000 tests with a clean output (no test timeouts).

```
jkirschner@HashiH0PW6D3:~/envoy$ bazel test --jobs=3 --runs_per_test=1000 //test/common/upstream:load_balancer_impl_test
INFO: Analyzed target //test/common/upstream:load_balancer_impl_test (0 packages loaded, 528 targets configured).
INFO: Found 1 test target...
[57 / 1,039] 3 actions running
[63 / 1,039] 3 actions running
[72 / 1,039] 3 actions running
[81 / 1,039] 3 actions running
[87 / 1,039] 3 actions running
[111 / 1,039] 3 actions running
[132 / 1,039] 3 actions running
[144 / 1,039] 3 actions running
[162 / 1,039] 3 actions running
[180 / 1,039] 3 actions running
[186 / 1,039] 3 actions running
[225 / 1,039] 3 actions running
[228 / 1,039] 3 actions running
[352 / 1,039] 3 actions running
[361 / 1,039] 3 actions running
[388 / 1,039] 3 actions running
[412 / 1,039] 3 actions running
[454 / 1,039] 3 actions running
[457 / 1,039] 3 actions running
[472 / 1,039] 3 actions running
[517 / 1,039] 3 actions running
[523 / 1,039] 3 actions running
[538 / 1,039] 3 actions running
[649 / 1,039] 3 actions running
[661 / 1,039] 3 actions running
[679 / 1,039] 3 actions running
[679 / 1,039] 3 actions running
[685 / 1,039] 3 actions running
[688 / 1,039] 3 actions running
[709 / 1,039] 3 actions running
[712 / 1,039] 3 actions running
[712 / 1,039] 3 actions running
[781 / 1,039] 3 actions running
[802 / 1,039] 3 actions running
[808 / 1,039] 3 actions running
[814 / 1,039] 3 actions running
[855 / 1,039] 3 actions running
[860 / 1,039] 3 actions running
[870 / 1,039] 3 actions running
[871 / 1,039] 3 actions running
[985 / 1,039] 3 actions running
[1,009 / 1,039] 3 actions running

TIMEOUT: //test/common/upstream:load_balancer_impl_test (Summary)
      /home/jkirschner/.cache/bazel/_bazel_jkirschner/bf0d63008780ca4fd369bad986d5988d/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/common/upstream/load_balancer_impl_test/run_13_of_1000/test.log

      [... lots of similar lines ...]
      /home/jkirschner/.cache/bazel/_bazel_jkirschner/bf0d63008780ca4fd369bad986d5988d/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/common/upstream/load_balancer_impl_test/run_615_of_1000/test.log
Target //test/common/upstream:load_balancer_impl_test up-to-date:
  bazel-bin/test/common/upstream/load_balancer_impl_test
INFO: Elapsed time: 81058.778s, Critical Path: 14339.84s
INFO: 1003 processes: 1050 linux-sandbox.
INFO: Build completed, 1 test FAILED, 1003 total actions
//test/common/upstream:load_balancer_impl_test                          TIMEOUT in 48 out of 1000 in 14298.8s
  Stats over 1000 runs: max = 14298.8s, min = 54.9s, avg = 241.2s, dev = 1007.5s
  /home/jkirschner/.cache/bazel/_bazel_jkirschner/bf0d63008780ca4fd369bad986d5988d/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/common/upstream/load_balancer_impl_test/run_13_of_1000/test.log

[... lots of similar lines ...]
  /home/jkirschner/.cache/bazel/_bazel_jkirschner/bf0d63008780ca4fd369bad986d5988d/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/common/upstream/load_balancer_impl_test/run_615_of_1000/test.log

Executed 1 out of 1 test: 1 fails locally.
```

Reply:
@jkirschner-hashicorp I can run it for you. I'll update the PR with results.

Reply:
Thanks @tonya11en. Given [your results](https://github.com/envoyproxy/envoy/pull/31507#issuecomment-1911240058) (10k test runs, all passing), I'll mark this as resovled.